ouroboros-network/test/Test/LedgerPeers.hs

```diff
@@ -0,0 +1,162 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE NamedFieldPuns    #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes        #-}
+
+module Test.LedgerPeers where
+
+import           Control.Exception (SomeException (..))
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadFork
+import           Control.Monad.Class.MonadSay
+import           Control.Monad.Class.MonadTime
+import           Control.Monad.Class.MonadThrow
+import           Control.Monad.IOSim
+import           Control.Tracer (showTracing, Tracer (..), traceWith)
+import           Data.List (foldl', intercalate)
+import           Data.List.NonEmpty (NonEmpty (..))
+import           Data.Word
+import           Data.Ratio
+import           System.Random
+
+import           Ouroboros.Network.PeerSelection.LedgerPeers
+
+import           Test.Tasty (TestTree, testGroup)
+import           Test.Tasty.QuickCheck (testProperty)
+import           Test.QuickCheck
+import           Text.Printf
+
+tests :: TestTree
+tests = testGroup "LedgerPeers"
+  [ testProperty "Pick 100%" prop_pick100
+  , testProperty "Pick" prop_pick
+  ]
+
+newtype ArbitraryRelayAddress = ArbitraryRelayAddress RelayAddress
+
+instance Arbitrary ArbitraryRelayAddress where
+    arbitrary = do
+      ArbitraryRelayAddress <$> elements [ RelayAddressAddr (read "1.1.1.1") 1234
+               , RelayAddressDomain (DomainAddress "relay.iohk.example" 1234)
+               ]
+
+data StakePool = StakePool {
+      spStake :: !Word64
+    , spRelay :: NonEmpty RelayAddress
+    } deriving Show
+
+
+
+instance Arbitrary StakePool where
+    arbitrary = do
+        stake <- choose (0, 1000000)
+        (ArbitraryRelayAddress firstRelay) <- arbitrary
+        moreRelays <- map unAddr <$> arbitrary
+        return $ StakePool stake (firstRelay :| moreRelays)
+      where
+        unAddr (ArbitraryRelayAddress a) = a
+
+newtype LedgerPools = LedgerPools [(PoolStake, NonEmpty RelayAddress)] deriving Show
+
+instance Arbitrary LedgerPools where
+    arbitrary = LedgerPools . calculateRelativeStake <$> arbitrary
+
+      where
+        calculateRelativeStake :: [StakePool] -> [(PoolStake, NonEmpty RelayAddress)]
+        calculateRelativeStake sps =
+            let totalStake = foldl' (\s p -> s + spStake p) 0 sps in
+            map (\p -> ( PoolStake (fromIntegral (spStake p) % fromIntegral totalStake)
+                       , spRelay p)) sps
+
+-- | A pool with 100% stake should allways be picked.
+prop_pick100 :: Word16
+             -> Property
+prop_pick100 seed =
+    let rng = mkStdGen $ fromIntegral seed
+        sps = [ (1, RelayAddressAddr (read "1.1.1.1") 1  :| [])
+              , (0, RelayAddressAddr (read "0.0.0.0") 0  :| [])
```

Comment:
if `RelayAddress` was abstract / polymorhpic over addresses & dns names, we wouldn't need to use real ip addresses here.

Reply:
I prefer using a concrete type in the test suite to having to add two extra parameters to `RelayAddress` in the code.